core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepositoryConnection.java

```diff
@@ -195,10 +199,23 @@ public SailQuery prepareQuery(QueryLanguage ql, String queryString, String baseU
 		ParsedQuery parsedQuery = QueryParserUtil.parseQuery(ql, queryString, baseURI);
 
 		if (parsedQuery instanceof ParsedTupleQuery) {
+			Optional<TupleExpr> sailTupleExpr = sailConnection.prepareQuery(ql, Query.Type.TUPLE, queryString, baseURI);
+			if (sailTupleExpr.isPresent()) {
+				parsedQuery = new ParsedTupleQuery(queryString, sailTupleExpr.get());
+			}
 			return new SailTupleQuery((ParsedTupleQuery) parsedQuery, this);
```

Comment:
Does this end up potentially parsing twice? Parsing queries is one of the slowest parts of the SHACL sail, so I think we should avoid parsing twice at all costs.

Reply:
I guess this is more of a legacy method, so we could probably live with having to parse the query twice if we maybe add some documentation to let the user know to prefer the more explicit methods.

Reply:
You're right that it parses twice, and of course that is a penalty. We could potentially improve it a bit by using some regex matching to determine the type (instead of a full parse) but at this point I didn't want to spend the effort.

Reply:
I'll add something to the javadoc to point to the more explicit methods though - it's a good thing to use those if you can, regardless.